pkg/execution/state/redis_state/lua/queue/enqueue.lua

```diff
@@ -69,6 +73,18 @@ if redis.call("HSETNX", queueKey, queueID, queueItem) == 0 and not is_normalize
   return 1
 end
 
+-- Check if the item is a singleton and if an existing item already exists
+if exists_without_ending(singletonKey, ":singleton:-") then 
+  if redis.call("EXISTS", singletonKey) ~= 0 then
+    return 2
+  end
+
+  -- Set the singleton key to the item ID
+  -- TODO: Remove the expiration when dequeuing the item removes the singleton key
+  redis.call("SETEX", singletonRunKey, 20, singletonKey)
+  redis.call("SETEX", singletonKey, 20, runID)
```

Comment:
If we set the keys before enqueuing non-atomically, there's a small window between setting the keys and enqueuing the item during which a failure or crash can occur. In such cases, the singleton key could be held indefinitely.